Vitamins, health, and nutrition

Xhuis opened the pull request:
:cl: Xhuis
add: Nutrition has been revamped! Food, drinks, and some medicines now have varying healthiness. Eating healthy foods and maintaining a good vitamin balance will allow you to very slowly heal minor wounds, and bolster your immune system!
add: Very slow natural healing has been added. If you're unhealthy, this will not function! Healing is roughly equal to 0.01 + (0.05% of your vitamin count). Your liver will also purge toxins at a rate of 0.05 + (0.02% of your vitamin count). Neither of these values can drop below zero.
add: Junk food like candy and soda will cause you to become unhealthy, whereas chef meals and plants will make you more healthy. Juices will also do this.
add: If a healthy diet isn't an option, vitamin supplements are available from the medical bay, as well as gamma globulin shots intended for malnutrition that massively increase vitamin count!
/:cl:

This adds a new health system; not health as in "how many hits can you take before dying" but "how many vitamins in your body" - basically physical wellness. This starts at zero, and when changed, will regress towards that neutral count at a rate that can't go below 0.05 per tick. 

Wellness is influenced in many different ways, with the primary one being diet. Sugary food and drinks like candy and soda will cause you to become unhealthy, while hot and fresh food from the chef will be healthier for you!

Healthy living has its benefits. This PR also adds very (*very*) slow natural healing to humans; this rate is different for each damage type:
* Brute: `0.01 + (vitamins * 0.0005)`
* Burn: `0.01 + (vitamins * 0.0003)`
* Toxins: `0.05 + (vitamins * 0.0002)`

At 40 vitamins, this would heal 0.03, 0.022, and 0.058 damage respectively, or 34, 46, and 18 ticks to heal one point of damage. these values *can* drop to zero through lack of proper eating, but they will never damage the user except in special cases.

Vitamins also play a hefty role in disease progression and immunity:
* Disease stage progression is now run against a check equal to `90 - vitamins`, which can dramatically slow how long it takes.
* Disease cure speed can be up to doubled with sufficient vitamins.
* If a check is passed equal to `0.66 * vitamins`, a disease might not be contracted.

Vitamin thresholds to be aware of:
* <=-10: Malnutrition
* 0: Neutral (vitamin count will always equalize to this)
* \>=25: Healthy
* \>=100: Hypervitaminitosis

Hypervitaminitosis is caused when the body has an excess of vitamins, and can usually only be reached with supplements. It will reverse the liver's natural toxin healing into toxin *damage* while above that threshold.

If a healthy diet isn't an option, or you simply have time-sensitive things to worry about, there are other options! The medical bay also stocks multivitamin tablets and gamma globulin shots, meant as a treatment for malnutrition. Multivitamin tablets increase healthiness by 15 per tablet but cannot exceed 5 healthiness total; gamma globulin shots provide a massive 75 boost and have no limit. Multivitamins are found copiously in NanoMeds, with gamma globulin shots also being stocked but in notably shorter supply.

Finally, multivitamins can also be created by heating a a 5/1 mixture of nutriment and vitamin to 400 K, or the same mixture with a 10/1 blood/vitamin count instead. Gamma globulin shots can be bought in the uplink but cannot be crafted due to their strength.

This addition encourages the chef to create good meals without shoehorning in minmaxing to determine the best one; any good that has vitamins in it will contribute to a healthy diet. Vitamin supplements, botany, mining, cargo, and science all having primary or secondary food sources also allows the crew to go without fear of mass malnutrition without a competent chef. It also doesn't punish those who want to use the more convenient vending machines, as regression towards neutral vitamin levels means that they won't be starved husks if they opt for the fastest option. For those that do care about the small boost in natural healing that eating well provides, it gives an incentive to interact with other crew, and makes cooking more relevant without forcing it to be a liability.

Thoughts, feedback, and input all appreciated!

Commit:
Health and nutrition system

Commit:
I did get these vending values, right?

optimumtact commented:
add: Very slow natural healing has been added

Hey man, did you not see the virus PR's?

Xhuis commented:
The healing in of itself is slow enough that it's more of a way to heal tiny amounts of damage without having to go to the medical bay for a papercut; by default it takes 34 ticks to heal 1 point of brute damage, and even that I think might be pressing it. It also comes at the caveat of eating and drinking well being its primary supplement, and without it it still won't be any good in a fight, with the main combat use being a point or two of healing between engagements while you lick your wounds. Even eating nutriment itself heals faster than that.

optimumtact commented:
good luck with the bikeshed design.

Xhuis commented:
If you could offer more complete thoughts, I'd be happy to take them into account.

Doppeltongue commented:
It's more like a way to heal tiny bits of damage you get by accident when someone throws something with force 1 at you, so that you aren't constantly bugged by that injured paperdoll and less than 100% health due to 1 damage.

QualityVan commented on code/modules/surgery/organs/stomach.dm:
```diff
@@ -48,6 +49,17 @@
+	//First, calculate the actual vitamin levels
+	var/vitamin_decay_rate = H.vitamins * 0.001 //Vitamin levels will slowly decay towards neutral
+	vitamin_decay_rate = Clamp(vitamin_decay_rate, -0.05, 0.05)
+	H.vitamins -= vitamin_decay_rate //Yes, this works for negatives, don't ask me how
```
it works for negatives because if h.vitamins is negative then VDR is negative and subtracting a negative pushes upward

Xhuis replied:
I spent at *least* an hour trying to figure that out. Ugh.

QualityVan commented on code/modules/reagents/chemistry/reagents/medicine_reagents.dm:
```diff
@@ -1121,3 +1121,24 @@
 	description = "A medication used to treat pain, fever, and inflammation, along with heart attacks."
 	color = "#F5F5F5"
+
+/datum/reagent/medicine/vitamin_supplement
```
It's a little weird that you're adding vitamin supplements when we already have vitamins as a reagent

Xhuis replied:
The vitamins reagent is just a "healthy" reagent; supplements make up for a lack of not having enough of it through consistent bad eating rather than not having it every few minutes.

Boredone commented:
Tox healing wise, there is an upgraded liver. Could have that heal a bit faster or something?

Xhuis commented:
I don't see why not.

vuonojenmustaturska commented:
hey, this is actually pretty cool

Xhuis commented:
You're pretty cool 😎

Hatterhat commented:
this seems fuckin radical

PKPenguin321 commented:
If passive healing becomes too much of a concern (unlikely with how small it is but w/e) you could change it so that it only works if you have 90 health or more so it only heals minor wounds. Maybe change the healing threshold up or down from 90 depending on how healthy you are

DapperMola commented:
So what chef foods will provide how much vitamins?

Xhuis commented:
Chef foods will probably end up giving between 0.6 and 3 healthiness depending on food. It decreases so slowly that the system takes consistent diet into account rather than short term nourishment. I might bump that up a bit, though.

Ratvar commented:
>medics eating all vitamines roundstart to get free slow healing
Why even medics have vitamines? "I eat garbage from vendors but also visit medbay so it's fin"

Repukan commented:
Making it slower on lizards

newfren commented:
Yo so as long as kevinz didn't lie to me it takes literally a minute and a half to heal a point of burn damage. There's actually no way this is relevant to combat at all. Good change.

XDTM commented:
Who would seriously powergame for 0.05 healing per second serously y'all hear regeneration and go crazy

Ratvar commented:
Make changeling power free cause it's so sucky?

XDTM commented:
Just give it passively to lings

Arianya commented:
Probably out of scope for this PR but it'd be cool if lings' nutrients were adjusted by the nutrient count of who they absorb. Gotta eat healthy, Thing!

Jalleo commented:
Can you make it take longer? 34 ticks is probably within a few seconds or a minute.

Grazyn commented:
>Making it slower on lizards

It's already more difficult for lizards, since fruits and vegetables (which are rich in vitamins) are poisonous to them.

Hatfish commented:
Fruits and vegetables are fine for lizards, it's dairy and grain that makes them sick.  Lizards should probably have slightly more regen since that's a common trait in most reptiles (shedding skin, regrowing tails, etc).

MoreRobustThanYou commented:
inb4 medbay raids the kitchen, steals all food, grinds it, and makes pills

MoreRobustThanYou commented:
@Xhuis maybe make it so nutrient count of people lings absorbed are also added to the "DNA profile"

RemieRichards commented:
@Arianya already exists

https://github.com/tgstation/tgstation/blob/70d0357b335736aaf27d7cc47968a27820d3f2d8/code/game/gamemodes/changeling/powers/absorb.dm#L58

TribeOfBeavers commented:
Is there any way for medbay to produce more vitamins or globulin shots?

Arianya commented:
@RemieRichards Wouldn't that not affect this, since that's nutrients and this is vitamins?

RemieRichards commented:
It wouldn't, but that's not what you asked.
>it'd be cool if lings' <B><i>nutrients</i></B> were adjusted by the <B><i>nutrient</i></B> count of who they absorb

Xhuis commented:
That changeling thing is actually a really neat touch that didn't occur to me. I'm not sure how I want to implement it but I *will* do something regarding it.

@Ratvar It can heal golems etc., yeah; currently all races consider vitamins the same way. I can probably adjust that based on stuff like food type. "ODing" on globulins if you have toxin immunity is a problem, so I'll probably make it halt healing while active.

Arianya commented:
Theorycrafting a bit more, it'd be neat if Changelings didn't get any vitamins from normal food/vitamin tablets/globulin, but we're instead pushed by an almost vampire-like need to feed.

Probably less fun in gameplay then in theory, but still

RemieRichards commented:
"eat this box of doughnuts"
"why are you still slow"

Arianya commented:
Thats specifically why I used vitamins rather then nutrients that time.

Vitamins adjust slowly over a trend of food, not from single bursts of food, per Xhuis explanation

Arianya commented:
Anyway what I'm now talking about is way out of scope but I'm just throwing around worthless ideas :woman_shrugging:

RemieRichards commented:
I will hammer your soul until there's no creativity left.

Arianya commented:
something something hammering me something something rule 8

RemieRichards commented:
you have a dirty mind.

Armhulen commented on code/game/gamemodes/changeling/powers/vitamin_synthesis.dm:
```diff
@@ -0,0 +1,18 @@
+//Vitamin Synthesis: Allows the changeling to shift their vitamin levels at will.
+/obj/effect/proc_holder/changeling/vitamins
+	name = "Vitamin Synthesis"
+	desc = "Allows us to shift our blood nutrient levels at will. Higher vitamin concentrations supplement the body's natural healing ability and allow faster recuperation of conventional wounds."
```
changeling power that give it vitamins, where it talks about the various health benefits of eating healthy

i'm dead

Commit:
Fixes, tweaks, changeling stuff

RemieRichards commented:
I'd request you remove the ling stuff, lest you make more work for myself.

Xhuis commented:
Are you working on something related to changelings?

Commit:
If I squash commits nobody will ever know my shame

RemieRichards commented:
I have an 80~% complete goonling port, have for a few days, I'm waiting on Antur's datum ling stuff before I continue.

Xhuis commented:
I can yank it out for now then, sure.

Xhuis commented:
* Removed changeling stuff at the request of Remie.
* Vitamin levels now affect resistance to disease, how easily it can be recovered from, and how quickly it worsens.
* Added a recipe for multivitamins by heating 5 parts nutriment and 1 part vitamin to 400 K, plus an alternate recipe that uses 10 parts blood instead. Gamma globulins, being rare and powerful, still have no recipe.
* Added gamma globulin shots to the uplink for 1 TC, with a max of two per uplink.

DoctoreMad commented:
>Gamma globulins, being rare and powerful, still have no recipe.
Added gamma globulin shots to the uplink for 1 TC, with a max of two per uplink.

Oi mate I think yer stuff just ain't that important.

Bawhoppen commented:
Just remove the auto-healing, any of it is bad

Xhuis commented:
If you'd have read the discussion, I've given multiple reasons why I think it's beneficial and adds more to the nutrition system than if it weren't there.

Bawhoppen commented:
I can still disagree?

Xhuis commented:
You're welcome to disagree, but don't state your opinion as fact.

Bawhoppen commented:
Are you really going to find fault in how I said something?

Xhuis commented:
"Just remove the auto-healing, any of it is bad"
Implies authority and makes you sound like an armchair game designer. So, yes, I got rustled by it.

QualityVan commented:
It now occurs to me that this doesn't touch satiation at all, meaning that there'll be two systems doing more or less the same thing in parallel, which seems kind of silly

Xhuis commented:
Satiation and vitamins are separate in real life; you can be sated but still not get enough vitamins to be healthy. Eating purely chips will still keep you full, but you'll be very unhealthy for it. Satiation here is the need to eat; vitamins instead determine how well the body and immune system function.

QualityVan commented:
look up mob's satiety var, it affects disease immunity, blood regeneration, and hunger increase rate, and is separate from hunger level. it's increased by vitamins and whatever's in cactus fruit

Xhuis commented:
I'd still stick with what I said; getting *enough* food is also important to all of those things, and not eating enough food will still make you vulnerable to disease. Both eating enough food and getting enough vitamins out of that food is important to function.

QualityVan commented:
both satiety and your new nutrition are systems that model 'getting enough vitamins', is my point
junk food lowers satiety and gives you penalties
healthy food increases your satiety and gives you bonuses

Xhuis commented:
Junk food doesn't negatively affect satiety; vitamins only affect it to a cap. They serve similar purposes but I think they both have niches all the same. Emphasizing this is the fact that low satiety makes you hungrier; I think it's coded as a short-term benefit from eating well rather than the long-term benefit vitamins provide.

QualityVan commented:
https://github.com/tgstation/tgstation/blob/af4d9a85c9a3eeef1e5e0637ef4ebd0312794ac7/code/modules/food_and_drinks/food/snacks.dm#L106

QualityVan commented:
Vitamins also only affect nutrition to a cap, that cap being 200, so I'm not sure what your point is there

Xhuis commented:
I shall argue this to the ends of the earth.

nicbn commented:
I think this is a really bad idea. Hypervitaminosis and subvitaminosis symptoms take days to manifest and subnutrition has nothing to do with vitamins as it is caused by the lack of lipids and sugar.
And human natural healing takes from hours to years.

Xhuis commented:
Like diseases advance extremely quickly so as to have an in-game relevance, vitamins and the immune system are the same way. It's realistic but much faster than in real life so that it's still relevant.

nicbn commented:
It actually makes sense that diseases advance extremely quickly. In space, you have almost no contact with microbies so you have less antibodies, however your digestive system stills work almost the same way, and also saying that X system is nonsense is not an excuse for introducing another unrealistic system.

nicbn commented:
Also saying that lack of vitamins cause malnutrition kills me inside out.

AutomaticFrenzy commented:
Haha, glad I saw this. I OD on vitamins in botany a lot of rounds

Commit:
Merge branch 'master' into health_revamp

Commit:
Merge branch 'master' into health_revamp

Commit:
Update snacks_bread.dm

Commit:
Updates for new deep fryer stuff

Xhuis commented:
This is ready!

MoreRobustThanYou commented:
btw @Xhuis why haven't you been on IRC in a long time?

Xhuis commented:
Just got bored of it. I don't really enjoy being there anymore, and when I *do* go there it's always vitriolic.

imsxz commented:
forgive me if someone mentioned this already, I havent read the entire discussion. might wanna have it check if a disease is harmful, otherwise it will slow down/prevent healing viruses.

Xhuis commented:
I would argue that the immune system can't really tell that. It detects foreign bacteria, and instinctually attacks them; it's why beneficial diseases take just as long and why satiety also prevents beneficial diseases in the same way. Even if it *does* help the body, you can't really tell white blood cells through a loudspeaker that the virus invading all your cells is a good thing.

optimumtact commented:
@tgstation/commit-access Do any of you actually want to merge this? I am going to close this in say 2 days unless one of you picks it up and gets it in.

MoreRobustThanYou commented:
i'm not a maintainer but i like this

nicbn commented:
I don't think we need features that aren't worth it to mantain.

Xhuis commented:
That argument doesn't make any sense...?

Commit:
Merge branch 'master' of https://github.com/tgstation/tgstation into health_revamp

Repukan commented:
Tune and tweak the same way you did with clockcult twice before getting burned out and leaving it as a burning heap?

Repukan commented:
"We already have a lot of unfinished or unmantained systems so one more wouldn't matter" is not a good argument either.

Xhuis commented:
Do you ever say anything with more than ten seconds of thought put into it?

Repukan commented:
Not an argument

Xhuis commented:
![](https://i.imgur.com/hmQtD78.png)

lzimann commented:
time's up

Xhuis commented:
You really gotta learn to stop holding a grudge, man.

lzimann commented:
not my fault no one cares about this, i'm sure a maintainer can re-open if they do

JJRcop commented:
You may want to consider splitting this up and adding it slowly with multiple PRs, it's a lot easier to review.

optimumtact commented:
they'd have to find some maintainer willing to review/merge it first
